### PR TITLE
[FLINK-12211][table-planner-blink] Add more it cases to blink batch

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Add BuiltInFunctionsITCase to test build in functions in complete SQL environment.
Add BinaryITCase to test sql binary types.
Add DateFunctionsITCase to test date time timestamp types.
Add StringITCase to test varchar type.
Add simple misc it cases.
Fix SqlFunctionUtils.lpad and rpad with pad is empty string.

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by wuchong** [comment]
The changes to lpad and rpad looks good to me. 

Btw, I would suggest to move tests in `BuiltInFunctionsITCase`, `DateFunctionsITCase`, `StringITCase` to unit tests under `org.apache.flink.table.expressions.utils`. Because integrated test will cost much more time than unit tests, and they can be covered by expression unit tests.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/batch/sql/DateFunctionsITCase.scala [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+      Seq(
+        row(UTCDate("2017-12-12")),
+        row(UTCDate("2017-12-12"))
+      ))
+  }
+
+  @Ignore
+  @Test
+  def testDateEqualsWithDateString(): Unit = {
+    // TODO: May be a bug
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/batch/sql/DateFunctionsITCase.scala]: create a jira for this?

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/batch/sql/DateFunctionsITCase.scala]: It is a legacy of the past, no longer a bug.

**Comment by JingsongLi** [comment]
> The changes to lpad and rpad looks good to me.
> 
> Btw, I would suggest to move tests in `BuiltInFunctionsITCase`, `DateFunctionsITCase`, `StringITCase` to unit tests under `org.apache.flink.table.expressions.utils`. Because integrated test will cost much more time than unit tests, and they can be covered by expression unit tests.

I think these ITCases are not just Expression tests, they are different from Expression tests in that they integrate CodeGen of sort of exprs into an Operator to achieve some different results.
Now with AbstractTestBase, tests are at the millisecond level and should not be slow.

**Comment by JingsongLi** [comment]
@wuchong Deduplicated cases and move some ITCases to expression test and Merge Class to Accelerate Testing

**Comment by JingsongLi** [comment]
https://github.com/apache/flink/pull/8165 make CalcITCase.testTimestampCompareWithDateString fail. Some bug in https://github.com/apache/flink/pull/8165 .

**Comment by godfreyhe** [comment]
> #8165 make CalcITCase.testTimestampCompareWithDateString fail. Some bug in #8165 .

The bug is in codegen, not in #8165. codegen needs to handle `Calc(select=[j], where=[<(j, CAST(_UTF-16LE'2017-11-11'))])` correctly. In blink, some advanced rules will convert `CAST(_UTF-16LE'2017-11-11'):TIMESTAMP(3) NOT NULL)` to a timestamp value directly, however those rules have not been port to flink yet.

**Comment by JingsongLi** [comment]
> > #8165 make CalcITCase.testTimestampCompareWithDateString fail. Some bug in #8165 .
> 
> The bug is in codegen, not in #8165. codegen needs to handle `Calc(select=[j], where=[<(j, CAST(_UTF-16LE'2017-11-11'))])` correctly. In blink, some advanced rules will convert `CAST(_UTF-16LE'2017-11-11'):TIMESTAMP(3) NOT NULL)` to a timestamp value directly, however those rules have not been ported to flink yet.

Because our CodeGen not support varchar(yyyy-mm-dd) cast to timestamp(return null). But calcite support it.
So if the plan use calcite method to reduce the cast expression, the case will past.
But if the plan did not reduce the cast, the case will fail.
I add todo to the case, Let's support later.

**Pull request merged**